speedy-links/src/speedy-links.ts

```diff
@@ -79,6 +79,10 @@ async function renderPage(url: string): Promise<void> {
   let module = modulesCache[urlWithoutHash];
   if (!module) {
     const moduleUrl = await opts.mapLinkUrlToModuleUrl(urlWithoutHash);
+    if (!moduleUrl) {
+      location.href = url;
+      return;
+    }
```

Comment:
I would be very surprised if you can fix it in another way since you are looking for something inside a mapping table. 
If there is no match I do not expect something else than a return undefined.
Hence I need to treat this return as a fallback value, hence my fix.

Reply:
second:
prevent possible errors in the module loading (due to a missing mapping or because of HTTP errors)
then I'd extend speedy-links `try-catch`

```ts
async function renderPage(url: string): Promise<void> {
  const urlWithoutHash = removeHash(url);
  try {
    let module = modulesCache[urlWithoutHash];
    if (!module) {
      // if not mapping or mapping is not dynamically imported for some reason, fallback to `location.href`
      const moduleUrl = await opts.mapLinkUrlToModuleUrl(urlWithoutHash);
      module = await import(moduleUrl);
      modulesCache[urlWithoutHash] = module;
    }
    if (opts.moduleExecutor) {
      await opts.moduleExecutor(module);
    } else {
      const render = module.default as () => Promise<void>;
      await render();
    }
  } catch (e) {
    location.href = url;
  }
}
```

Reply:
any other use cases you had in mind?

Reply:
> second: prevent possible errors in the module loading (due to a missing mapping or because of HTTP errors) then I'd extend speedy-links `try-catch`
> 
> ```ts
> async function renderPage(url: string): Promise<void> {
>   const urlWithoutHash = removeHash(url);
>   try {
>     let module = modulesCache[urlWithoutHash];
>     if (!module) {
>       // if not mapping or mapping is not dynamically imported for some reason, fallback to `location.href`
>       const moduleUrl = await opts.mapLinkUrlToModuleUrl(urlWithoutHash);
>       module = await import(moduleUrl);
>       modulesCache[urlWithoutHash] = module;
>     }
>     if (opts.moduleExecutor) {
>       await opts.moduleExecutor(module);
>     } else {
>       const render = module.default as () => Promise<void>;
>       await render();
>     }
>   } catch (e) {
>     location.href = url;
>   }
> }
> ```

This one is causing either an infinite loop in the render because of modification of the popstate inside the render.
Or it fetches and possible missing module causing a 404.

Reply:
as previously discussed in Discord: I couldn't reproduce an infinite loop error

as for 404, yes, it's good idea to prevent it too, so my solution can be impoved as well, but it's just a conole.error which doesn't impact the UI and eventually `catch` fallback will take over it

the reason I want to extend try-catch is that `mapLinkUrlToModuleUrl` function might contain some bugs as well and it's better to fallback to the `location.href` instead of raising an exception because of that function and stopping navigation

so you can leave your solution as well, but consider improving it to solve this situation as well
or improve mine to solve 404 console error (prevent unnecessary dynamic import)